BarrysPeanutsTemporalJava/src/main/java/barryspeanuts/ShoppingCartWorkflowImpl.java

```diff
@@ -0,0 +1,98 @@
+package barryspeanuts;
+
+import barryspeanuts.model.Address;
+import barryspeanuts.model.CreditCard;
+import barryspeanuts.model.PurchaseItem;
+import io.temporal.workflow.Workflow;
+import io.temporal.workflow.WorkflowQueue;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.json.JSONObject;
+import org.slf4j.Logger;
+
+public class ShoppingCartWorkflowImpl implements ShoppingCartWorkflow {
+  private static final Logger logger = Workflow.getLogger(ShoppingCartWorkflowImpl.class);
+
+  boolean exit = false;
+  List<PurchaseItem> purchaseItems = new ArrayList<>();
+  private final WorkflowQueue<Runnable> queue = Workflow.newWorkflowQueue(1024);
+
+  @Override
+  public void startWorkflow() {
+    logger.info("Starting Workflow for Barry's Peanuts");
+    Workflow.await(() -> exit);
+  }
+
+  @Override
+  public List<PurchaseItem> queryPurchaseItems() {
+    return this.purchaseItems;
+  }
+
+  @Override
+  public void addItem(PurchaseItem purchaseItem) {
+    this.purchaseItems.add(purchaseItem);
+  }
+
+  @Override
+  public void addItems(List<PurchaseItem> purchaseItems) {
+    this.purchaseItems.addAll(purchaseItems);
+  }
+
+  @Override
+  public void removeItem(PurchaseItem purchaseItem) {
+    this.purchaseItems.remove(purchaseItem);
+  }
+
+  @Override
+  public void checkOut() {
+    logger.info("Checking out {} purchase items.", this.purchaseItems.toArray().length);
+  }
+
+  @Override
+  public void pay(Address billingAddress, CreditCard creditCard) {
+    for (PurchaseItem purchaseItem : this.purchaseItems) {
+      purchaseItem.setBillingAddress(Optional.ofNullable(billingAddress));
+    }
+
+    JSONObject jsonObj = new JSONObject(billingAddress);
+    logger.info(
+        "Paying for purchase using credit card number {} for {} purchase item at billing address {}.",
+        creditCard.getNumber(),
+        this.purchaseItems.toArray().length,
+        jsonObj);
+  }
+
+  @Override
+  public void ship(Address shippingAddress, String shipper) {
+    for (PurchaseItem purchaseItem : this.purchaseItems) {
+      purchaseItem.setShippingAddress(Optional.ofNullable(shippingAddress));
+    }
```

Comment:
Same as pay, I don't think you should mutate purchase items. Just leave them immutable. If you need to keep this address overall, can do.

Reply:
Please see my comment above.

Reply:
Addressed in that I have refactored code to use the `CheckoutInfo` class as a parameter to the `checkout()` function.